src/EventStore.Projections.Core/Services/Interpreted/JintProjectionStateHandler.cs

```diff
@@ -993,9 +987,266 @@ private bool EnsureLinkMetadata(out JsValue value)
+				return _bufferWriter.WrittenMemory;
+			}
+
+			void PushWriter(IWriter writer) {
+				if (_depth == _iterators.Length - 1) {
+					throw new InvalidOperationException("Depth exceeded");
+				}
+				_iterators[++_depth] = writer;
+			}
+
+			bool PopWriter() {
+				_iterators[_depth] = _completed;
+				_depth--;
+				return _depth >= 0;
+			}
+
+			interface IWriter {
+				bool Write(Utf8JsonWriter writer);
+			}
+			class ArrayWriter : IWriter {
+				private int _position;
+				private int _length;
+				private ArrayInstance _instance;
+				private readonly Serializer _parent;
+
+				public ArrayWriter(ArrayInstance instance, Serializer parent) {
+					_position = -1;
+					_length = (int)instance.Length;
+					_instance = instance;
+					_parent = parent;
+				}
+
+				public bool Write(Utf8JsonWriter writer) {
+					if (_position == -1) {
+						writer.WriteStartArray();
+						_position++;
+					}
+
+					for (; _position < _length; _position++) {
+						var current = _instance[(uint)_position];
+						if (current.Type == Types.Object) {
+							if (current is ArrayInstance ai) {
+								_parent.PushWriter(new ArrayWriter(ai, _parent));
+								_position++;
+								return true;
+							} else {
+								_parent.PushWriter(new ObjectWriter(current.AsObject(), _parent));
+								_position++;
+								return true;
+							}
+						}
+						SerializePrimitive(current, writer);
+					}
+					writer.WriteEndArray();
+					return _parent.PopWriter();
+				}
+			}
+
+			class ObjectWriter : IWriter {
+				private readonly IEnumerator<KeyValuePair<JsValue, PropertyDescriptor>> _iterator;
+				private readonly Serializer _parent;
+				private bool _started;
+
+				public ObjectWriter(ObjectInstance instance, Serializer parent) {
+					_iterator = instance.GetOwnProperties().GetEnumerator();
+					_parent = parent;
+				}
+
+				public bool Write(Utf8JsonWriter writer) {
+					if (!_started) {
+						writer.WriteStartObject();
+						_started = true;
+					}
+					while (_iterator.MoveNext()) {
+						var (name, propertyDescriptor) = _iterator.Current;
+						var value = propertyDescriptor.Value;
+						_parent.WriteMaybeCachedPropertyName(name.AsString());
+						if (value.Type == Types.Object) {
+							if (value is ArrayInstance ai) {
+								_parent.PushWriter(new ArrayWriter(ai, _parent));
+								return true;
+							} else {
+								_parent.PushWriter(new ObjectWriter(value.AsObject(), _parent));
+								return true;
+							}
+						} else {
+							SerializePrimitive(value, writer);
+						}
+
+					}
+
+					writer.WriteEndObject();
+					return _parent.PopWriter();
+				}
+			}
+
+			[MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+			private void WriteMaybeCachedPropertyName(string name) {
+				if (_knownPropertyNames.TryGetValue(name, out var propertyName)) {
+					_writer.WritePropertyName(propertyName);
+					return;
+				}
+				if (_knownPropertyNames.Count < 1000) {
```

Comment:
just want to check that it's intentional that this will remember the first 1000 property names across the lifetime of the node

Reply:
it's for the lifetime of the projection and tied to the projection